[Console] Handle encoded characters in API requests

mibragimov opened the pull request:
Closes https://github.com/elastic/kibana/issues/131798

Fixes sending requests with encoded characters to ES. To test this out, issue the following requests:

1. Create some docs with `/` in ID:
```
POST _bulk
{ "index" : { "_index" : "my-index", "_id" : "this/is/a/doc" } }
{ "title" : "a doc" }
```

2. Retrieve the document:

```
GET my-index/_doc/this%2Fis%2Fa%2Fdoc
```
<img width="1500" alt="Screen Shot 2022-05-13 at 13 58 12" src="https://user-images.githubusercontent.com/53621505/168249079-23c780d1-3729-4593-8fa5-98b1ad06df1f.png">

To verify it works with different formats, issue the following requests:

```
PUT %{[@metadata][beat]}-%{[@metadata][version]}-2020.08.23
```

```
PUT /%3Cmy-index-%7Bnow%2Fd%7D%3E
```
<img width="1500" alt="Screen Shot 2022-05-13 at 14 01 07" src="https://user-images.githubusercontent.com/53621505/168249614-565fa068-38b5-4099-8c38-523809396cb2.png">

<img width="1500" alt="Screen Shot 2022-05-13 at 14 01 32" src="https://user-images.githubusercontent.com/53621505/168249683-1cc696e9-507d-4cd9-a5e8-c4c7ae613491.png">

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

mibragimov commented:
@elasticmachine merge upstream

elastichelix commented on src/plugins/console/server/lib/proxy_request.ts:
```diff
@@ -22,6 +22,7 @@ interface Args {
   timeout: number;
   headers: http.OutgoingHttpHeaders;
   rejectUnauthorized?: boolean;
+  path: string;
```
Why is this required? What is the impact on consumers with this addition?

elastichelix commented on src/plugins/console/server/lib/proxy_request.ts:
```diff
@@ -58,11 +54,18 @@ export const proxyRequest = ({
   timeout,
   payload,
   rejectUnauthorized,
+  path,
```
It is not immediately clear to me what the difference between `path` and `pathname` are. Can you use more specific names to show how they are different?

elastichelix commented on src/plugins/console/server/lib/proxy_request.ts:
```diff
@@ -58,11 +54,18 @@ export const proxyRequest = ({
+  let pathname = uri.pathname;
   let resolved = false;
+  // Compare original request path to percent-encoded path returned by node URL method
+  const requiresEncoding = trimStart(path, '/') !== trimStart(pathname, '/');
```
This solution seems specific to the the `/` character. Could there be other cases where the pathname requires encoding? What about other characters covered here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

elastichelix replied:
Ah ok sorry I didn't get this on first read through for some reason. This LGTM

mibragimov commented:
👋  @elastichelix, Thank you for reviewing this PR. I've addressed your feedback with [6d35a3d](https://github.com/elastic/kibana/pull/132191/commits/6d35a3d8408836da3f12802a5598e542e214e80e). Let me know what you think

mibragimov commented:
@elasticmachine merge upstream

elastichelix reviewed (approved):
LGTM

mibragimov commented:
@elasticmachine merge upstream

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.3|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/133601">](https://github.com/elastic/kibana/pull/133601)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
<!--- Backport version: 7.4.0 -->
